File: Code/Avernakis Nodejs/Avernakis-Nodejs/Private/UiRibbonTab.cpp
#include "StdAfx.h"
#include "UiRibbonTab.h"
#include "UiRibbonGroup.h"

#define ThisMethod($x) &UiRibbonTab::$x
#define AutoAddMethod($x, ...) AddMethod<__VA_ARGS__>( #$x, ThisMethod( $x ) )
#define MakeThisFunc($x) MakeFunc( this, &UiRibbonTab::$x )

namespace Nav
{

	namespace
	{
		ObjectRegister<UiRibbonTab> c_obj;
	}

	void UiRibbonTab::DefineControl()
	{
		AutoAddMethod( SetText );
		AutoAddMethod( GetText );
		AutoAddMethod( GroupGetCount );
		AutoAddMethod( GroupAdd );
		AutoAddMethod( GroupInsert );
		AutoAddMethod( GroupGetByIndex );
		AutoAddMethod( GroupGetById );
		AutoAddMethod( GroupRemoveByIndex );
		AutoAddMethod( GroupRemoveById );
		AutoAddMethod( GroupRemove );
		AutoAddMethod( GroupRemoveAll );
	}

	U1 UiRibbonTab::Ctor( UiWindow * p, Napi::Value v )
	{
		if ( !__CreateControl( p, v ) )
			return false;

		return true;
	}

	U1 UiRibbonTab::GroupAdd( UiRibbonGroup * pChild )
	{
		return nullptr != GetControlTyped().GroupAdd( Ui::RibbonGroup( (Ui::IRibbonGroup*) pChild->TakeOwnership().Detach() ) );
	}

	U1 UiRibbonTab::GroupInsert( U32 nInsertBefore, UiRibbonGroup * pChild )
	{
		return nullptr != GetControlTyped().GroupInsert( nInsertBefore, Ui::RibbonGroup( (Ui::IRibbonGroup*) pChild->TakeOwnership().Detach() ) );
	}

	UiRibbonGroup * UiRibbonTab::GroupRemove( UiRibbonGroup * pChild )
	{
		return (UiRibbonGroup*) (UiControl*) __ChangeContent( (Ui::Control&&) GetControlTyped().GroupRemove( (Ui::IRibbonGroup*) &pChild->GetControl() ) );
	}

}

File: Code/Avernakis Nodejs/Test-Nodejs/src/Ave/Ui/Ribbon/UiRibbonTab.ts
import { IControl } from "../UiControl";
import { AveLib } from "../../AveLib";
import { WindowLike } from "../Control/UiWindow";
import { RibbonGroup } from "./UiRibbonGroup";

export interface IRibbonTab extends IControl {
	new (window: WindowLike, key?: string | number): IRibbonTab;

	SetText(s: string): RibbonTab;
	GetText(): string;

	GroupGetCount(): number;
	GroupAdd(pChild: RibbonGroup): boolean;
	GroupInsert(nInsertBefore: number, pChild: RibbonGroup): boolean;
	GroupGetByIndex(nIndex: number): RibbonGroup;
	GroupGetById(nId: number): RibbonGroup;
	GroupRemoveByIndex(nIndex: number): RibbonGroup;
	GroupRemoveById(nId: number): RibbonGroup;
	GroupRemove(pChild: RibbonGroup): RibbonGroup;
	GroupRemoveAll(): RibbonTab;
}

export class RibbonTab extends (AveLib.UiRibbonTab as IRibbonTab) {
	private m_Group: Set<RibbonGroup> = new Set();

	GroupAdd(pChild: RibbonGroup) {
		this.m_Group.add(pChild);
		return super.GroupAdd(pChild);
	}

	GroupInsert(nInsertBefore: number, pChild: RibbonGroup) {
		this.m_Group.add(pChild);
		return super.GroupInsert(nInsertBefore, pChild);
	}

	GroupRemoveByIndex(nIndex: number) {
		const t = super.GroupRemoveByIndex(nIndex);
		this.m_Group.delete(t);
		return t;
	}

	GroupRemoveById(nId: number) {
		const t = super.GroupRemoveById(nId);
		this.m_Group.delete(t);
		return t;
	}

	GroupRemove(pChild: RibbonGroup) {
		this.m_Group.delete(pChild);
		return super.GroupRemove(pChild);
	}

	GroupRemoveAll() {
		this.m_Group.clear();
		return super.GroupRemoveAll();
	}
}
